### PR TITLE
[Conformance checking] More refactoring of associated type inference.

### PR DESCRIPTION
**Pull request by DougGregor**
Move the rest of associated type inference into the new source file,
and split RequirementEnvironment into its own AST-level header, because it
can be re-used and has no ties to the type checker.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: include/swift/AST/RequirementEnvironment.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
+//===--- RequirementEnvironment.h - Swift Language Type ASTs ----*- C++ -*-===//
```
Comment (slavapestov) [inline include/swift/AST/RequirementEnvironment.h]: Wrong description here

File: lib/AST/RequirementEnvironment.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
+//===--- RequirementEnvironment.cpp - ASTContext Implementation -----------===//
```
Comment (slavapestov) [inline lib/AST/RequirementEnvironment.cpp]: Also wrong description here

File: include/swift/AST/RequirementEnvironment.h (hunk 1 above)
Comment (DougGregor) [reply include/swift/AST/RequirementEnvironment.h]: Drat, okay. Thanks!

File: lib/AST/RequirementEnvironment.cpp (hunk 2 above)
Comment (DougGregor) [reply lib/AST/RequirementEnvironment.cpp]: (FWIW, I fixed these in a subsequent, completely-unrelated PR)

**Pull request merged**